Add tumor multi

tru0067 opened the pull request:
Tested and working properly on tumors! I haven't fully tested how it behaves with every other possible effect/trait, but since it is just using the same process as e.g. `DamageMod` I imagine it should be fine.

One note is that since it is a multiplier applied to the damage itself, it does mean that it interacts in a different way than the Precision Multiplier on enemies with a low Weakspot Multiplier (in the base game this is really only the Nightmare Scout - base game Scattergun can kill Nightmare Scout regardless of the Precision Multiplier, but a low tumor multi might prevent it from doing so).
- This could be changed so that it instead modified the Precision Multiplier of the hit.
- I believe this would also change how it interacts with other effects/traits which use the damage ignoring precision.

### Context
Used `WeaponDamageContext` (I think we discussed using `WeaponPreHitEnemyContext`) as this seems to be where most multipliers are applied.

### Naming Scheme
I went with `Tumor/TumorMulti.cs` with `TumorMulti.Multi` as the relevant number. This is unlike the `Mod` naming elsewhere in the code, but to me a potentially stacking modifier and a flat multi seemed like distinct things. Easy to change if you want `TumorMod.Mod` instead.

tru0067 committed:
Add tumor multi

tru0067 commented:
If you're happy with everything I'm also happy to clone the wiki and add things there as well, then you can copy pasta from there (no PRs for wikis unfortunately).

Dinorush commented:
Looks mostly good, but there are a couple things that could be changed:
1. As this cannot stack, it should be a trait, not an effect. I really should've enforced that via inheritance but I just didn't think of implementing it at the time. I can do that later, will likely just be an additional abstract class that defines the `AllowStack` variable so it's obvious from the inheritance tree rather than being defined in each individual effect/trait.
2. (Optional) The behavior of this is evidently intended to match `PrecisionDamageMulti`, but the naming scheme differs (`TumorMulti` class name and `multi` field). I think either renaming it to `TumorDamageMulti` OR adding `tumordamagemulti` and `tumormulti` to the accepted deserialize list would be best.

Dinorush commented:
Oh, and regarding the other things you mentioned:
- This runs before damage is actually dealt to the enemy, so even if tumor multiplier is 0, it will still deal body shot damage at minimum due to damage calculations (same thing that occurs if precision multiplier is 0).
- None of the other effects ignore precision, but the user can simply set precision multiplier to 0. The tumor multiplier is basically just a multiplier on the precision multiplier, so in the end it will effectively be 0 in such a case too. The only concern is that this will not apply to DoT and Explosive effects that have `IgnoreDamageMods` enabled, but that is an exceedingly rare use case. There is no need to accommodate it, just note it on wiki. 
- We discussed the PreHit contexts for the piercing multiplier (since that would need access to the ray data IIRC). Tumor multiplier is indeed fine with the Damage context alone.

tru0067 committed:
Change from effect to trait

tru0067 commented:
Have now changed it to a trait!

> 2. (Optional) The behavior of this is evidently intended to match `PrecisionDamageMulti`, but the naming scheme differs (`TumorMulti` class name and `multi` field). I think either renaming it to `TumorDamageMulti` OR adding `tumordamagemulti` and `tumormulti` to the accepted deserialize list would be best.

`WeaponHitData` uses `precisionMulti` and `DamageOverTime` and `Explosive` use `PrecisionMult` with `precisionmult` and `precision` for deserializing, so I'm not sure if `TumorDamageMulti` actually fits with anything else.

Perhaps renaming to `Tumor/TumorMod`, the relevant number being `TumorMult` and with `tumormult` and `tumor` for deserializing would be the best fit with everything else? Also leaves room for other tumor modifiers.

tru0067 commented:
> * This runs before damage is actually dealt to the enemy, so even if tumor multiplier is 0, it will still deal body shot damage at minimum due to damage calculations (same thing that occurs if precision multiplier is 0).

It is definitely possible to reduce the damage below what would be possible with just precision multi. 0.01 tumor multi sniper deals 4 damage to Mothers, while a 0.02 precision multi sniper would still deal 40.

Dinorush commented:
While the naming scheme should remain consistent within the mod, ideally it is also similar to vanilla or intuitive so rundown developers have to do less lookups. To that end I think it would be better to rename the other fields to PrecisionDamageMulti/StaggerDamageMulti and expand their accepted deserialize list, then adjust TumorMulti to match, rather than keep it as-is since TumorMulti doesn't match anything atm. I can handle modifying the others myself.

I don't think it best to name it with the `Mod` suffix since it differs greatly from them. While, yes, it does modify a weapon value like the others, the `Mod` suffix is currently limited to timed effects that are triggered and can stack multiple times. AmmoCap and Accelerate do not have mod in the name despite modifying weapon values to reflect this difference between effects and traits.

Regarding damage, yeah I thought through that incorrectly. You're right, since the the damage minimum is merely comparing precision * weakspot to be greater than the damage given which is what we are modifying directly, nothing is keeping the tumor damage above minimum. In that case, you could either
1. Add a context that changes precision multi (as you suggested earlier). Would need to cache precision multi to reset it on pierce like damage is currently done. Requires no further changes though and would be compatible with Explosive/DoT regardless of their `IgnoreDamageMods` field.
2. Calculate the correct multiplier on Tumor damage so that it deals body shot damage at minimum. Requires less change from what you have but preserves the edge case issue with DoT/Explosive.

I would suggest 1) since that seems to be a more robust approach. I don't believe we should have any option that supports damage being less than non-precision damage since nothing else does that.

Dinorush commented:
Actually, 1) would be a little more complex than that. You'd also have to call the context on DoT and Explosive where the damage is calculated. So it would require a little more effort, certainly more than 2). I'll leave the call up to you, I'm fine with either approach since the code bloat of 1) is still rather small.

tru0067 committed:
Rename field and change deserialize list

Field `Multi` -> `TumorDamageMulti` and deserialize list ("multiplier",
"multi") -> ("tumordamagemulti", "tumormulti")

tru0067 commented:
Renames done, now `TumorMulti.TumorDamageMulti` with `tumordamagemulti` and `tumormulti` as possible deserializations.

I think I'd rather go with option 1. Option 2 seems like it might run into a lot of edge-cases.

One way to go about doing option 1 that appeals to me is to essentially refactor out `WeaponStackModContext` and reintroduce its functionality as a new `StackMod` type. `WeaponDamageContext`, `WeaponFireRateContext` and `WeaponRecoilContext` would all then have a new `StackMod` field.

`StackMod` would implement `Value` and `AddMod` in the same way that `WeaponStackModContext` currently does. Accessing one of these context values would then change from something like `context.Value` to `context.Damage.Value` or `context.Recoil.Value` etc..

Then `WeaponDamageContext` could simply have two of these `StackMod` fields. `context.Damage` and `context.PrecisionDamageMultiplier`. `TumorMulti` would modify the later. The weaponRayData would then get reset with these two values in the weapon patch.

This would also make it simple to later add some other fun effects, like stacking stagger multipliers etc.

tru0067 committed:
Fix change to trait not being reflected elsewhere

Dinorush commented:
I agree, it would be best to separate the StackMod functionality so we can have multiple in the same context. I'll accept the PR now and make the StackMod later.
